### PR TITLE
DEP: spatial.distance: deprecate `kulczynski1` and `sokalmichener` metrics

### LINKED ISSUE #21187: BUG: spatial.distance: `kulczynski1` returns similarity rather than dissimilarity
gh-2009 reported wrong formula for Kulczynski distance. gh-14588 fixed it by adding a kulczynski1 function, according to a formula in [Sh93] (Table 1). The same formula appears in [DD16] (p. 305).  The original Kulczynski function was deprecated and later removed.

I think the current implementation has a bug in that it returns the Kulczynski1 *similarity*, rather than *dissimilarity*. This is evident from [Sh93], whose Table is titled “similarity coefficients”. Another way to realize this is to note that the result has $n_{TT}$ in the numerator, i.e. the more matching ones, the larger return value (opposite to what a distance does).

On the other hand, no issue appears to have been reported related to usage of kulczynski1. If it is used in for example clustering and returns similarity rather than distance, I’d expect a user to find it out very quickly. That no one reported, combined with the historical difficulty in locating a definitive source for the formula, suggests to me that the function is not (widely) used.

Therefore, I’m inclined to suggest that the function be deprecated and removed altogether from spatial. There is no real loss of functionality because the current implementation can be reproduced by $(1-J)/J$ where $J$ is the Jaccard distance.

### Further reading

Let me recap the "kulczynski1" distance definitions I've seen so far. Given two boolean vectors, let $a$ denote the number of matching ones, $b$ denote the number of 0-1, $c$ denote the number of 1-0, and $d$ denote the number of matching zeros.

1. The current [kulczynski1 function](https://github.com/scipy/scipy/blob/92d2a8592782ee19a1161d0bf3fc2241ba78bb63/scipy/spatial/distance.py#L829-L894) computes

$$\frac{a}{b+c}$$

This is referred to as "Kulczynski unnamed 1" binary similarity coefficient in [Sh93].

2. The deprecated and removed [kulsinski function](https://github.com/scipy/scipy/blob/44e31b1eb202ef91caad059952ba268d2610a60b/scipy/spatial/distance.py#L878) computes

$$
1-\frac{a}{a+2(b+c)+d}
$$

3. A [kulczynski (dissimilarity) index](https://mothur.org/wiki/kulczynski/) referenced in [this comment](https://github.com/scipy/scipy/issues/2009#issuecomment-17026788) computes

$$
1-\frac{a}{b+c}
$$

4. A [kulczynski-cody (dissimilarity) index](https://mothur.org/wiki/kulczynskicody/) referenced from the same comment computes

$$1-\frac{1}{2}\left(\frac{a}{a+b}+\frac{a}{a+c}\right)$$

One minus this index is referred to as "Kulczynski unnamed 2" binary similarity coefficient in [Sh93]. The discussions in gh-2009 appears to suggest this definition is the more popular one.

5. A "Kulczynski coefficient" referenced in [KR90], page 65 computes

$$\frac{d}{b+c}$$

### References

[DD16] Deza, M. M and Deza, E. (2016). Encyclopedia of Distances.

[KR90] Kaufman, L. and Rousseeuw, P. J.  (1990).  "Finding Groups in Data: An Introduction to Cluster Analysis."  John Wiley & Sons, Inc.

[Sh93] Shi, G. R. (1993). “Multivariate data analysis in palaeoecology and palaeobiogeography—a review. *Palaeogeography, Palaeoclimatology, Palaeoecology*, 105(3-4): 199-234.

### PR DESCRIPTION
**Pull request by fancidev**
#### Reference issue
Closes gh-21187, gh-2011.

#### What does this implement/fix?
This PR deprecates two metrics under `scipy.spatial.distance` in order to clean up the namespace.

*kulczynski1*. As explained in gh-21187, the `kulczynski1` metric is documented to be a *dissimilarity* measure, as is all the rest metrics, but its return value is in fact a *similarity* measure. The fact that no one has complained about this suggests that this metric might not actually be in use. Therefore this PR deprecates the `kulczynski1` metric and schedules it for removal later.

*sokalmichener*. As explained in gh-2011, `sokalmichener` returns the same as `rogerstanimoto`, making (the implementation of) one of them redundant. In addition, in the literature the Sokal-Michener dissimilarity almost always refers to the `matching` similarity, which does not agree with the current implementation of `sokalmichener`. Therefore this PR deprecates the `sokalmichener` metric and schedules for removal later.

#### Additional information
I considered three solutions to resolve the contradiction between the documentation and implementation of `kulcyznski1`:

1. Amend the doc to state that the metric returns a *similarity* index. Pro: backward compatible. Con: different from all the other distance metrics, and probably not what the function is intended for.
2. Modify the implementation to return a *dissimilarity* index. Pro: consistent with doc and other functions. Con: "flipping" the return value (even with deprecation warning or future warning) breaks backward compatibility in a subtle way that could lead to serious bugs.
3. Deprecate and later remove the `kulczynski1` metric altogether. Pro: cleaner code base, less future maintenance. Con: existing code that does rely on `kulczynski1` returning a similarity index has to be updated to use an equivalent formula, e.g. `0.5*(1/D-1)` where `D` is the [dice distance](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.dice.html#scipy.spatial.distance.dice).

This PR goes with option 3.

Similar rationale applies to the `sokalmichener` metric.

### CODE REVIEW

**Comment by lucascolley** [comment]
cc @WarrenWeckesser you commented about the formula for this in https://github.com/scipy/scipy/issues/2009#issuecomment-17026788, and @czgdp1807 you added the function.

**Comment by fancidev** [comment]
The CI errors are "new" with pytest 8.3.3 -- the tests pass with pytest 7.4.4. Looking into that now.

**Comment by fancidev** [comment]
Let me recap the few "kulczynski1" distance definitions seen out there. Given two boolean vectors, follow the notations of [KR90] and let $a$ denote the number of matching ones, $b$ denote the number of 0-1, $c$ denote the number of 1-0, and $d$ denote the number of matching zeros.

1. The current `kulczynski1` function computes

$$\frac{a}{b+c}$$

2. A [kulczynski1 index](https://mothur.org/wiki/kulczynski/) referenced from @WarrenWeckesser's https://github.com/scipy/scipy/issues/2009#issuecomment-17026788 computes

$$
1-\frac{a}{b+c-2a}
$$

3. A [kulczynski-cody index](https://mothur.org/wiki/kulczynskicody/) referenced from the same comment computes

$$1-\frac{1}{2}\left(\frac{a}{b}+\frac{a}{c}\right)$$

4. A "Kulczynski coefficient" referenced in [KR90], page 65 computes

$$\frac{d}{b+c}$$

**References**

[KR90] Kaufman, L. and Rousseeuw, P. J.  (1990).  "Finding Groups in Data: An Introduction to Cluster Analysis."  John Wiley & Sons, Inc.

**Comment by fancidev** [comment]
@tylerjereddy Could you take a look at this PR when you have time? Part of it would close a ten-year-old (!) issue (gh-2011). I’m hoping to make it into the 1.15 release to kick off the deprecation process sooner than later!

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -1573,8 +1591,20 @@ def russellrao(u, v, w=None):
     return float(n - ntt) / n
 
 
+def _deprecate_sokalmichener(fn):
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        warnings.warn(
+            "The sokalmichener metric is deprecated since Scipy 1.15.0 and will "
+            "be removed in SciPy 1.17.0.", DeprecationWarning, stacklevel=2)
```
Comment (tylerjereddy) [inline scipy/spatial/distance.py:1599]: should we mention the metric folks should use instead?

**Review by tylerjereddy — COMMENTED** [review]
I added some initial comments. The case for deprecating `sokalmichener` seems more straightforward to me, so might make sense to just do that one to start; I didn't see any comments/responses for `kulczynski1` in gh-21187, which makes me a bit more nervous.

I tried doing some GitHub code searching to see how popular these metrics are, but didn't reach a decisive conclusion. It may make sense to ask i.e., CJ to do a quick check on the Google code search to see if they are popular.

One other thing I haven't seen discussed is that `scikit-learn` also has distance computation infrastructure quite similar to ours: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.DistanceMetric.html

They have `sokalmichener` and `rogerstanimoto` both defined identically in their docs, so I'd like to see some discussion/coordination with them on:

1) why they are doing that? (because of us?)
2) if they agree on the decision to have only one of them moving forward (we don't need their permission, but it might be good to just check with them)

They also have `kulsinski` defined there, so I'd again be interested in hearing their thoughts/coordinating a little on it. Maybe a quick discourse thread to coordinate a bit, or just ping them here maybe..

**Comment by tylerjereddy** [comment]
cc @thomasjpfan on sklearn side perhaps

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -949,6 +963,10 @@ def kulczynski1(u, v, *, w=None):
     -3.0
 
     """
+    warnings.warn(
+        "This function is deprecated and will be removed in SciPy 1.17.0.",
+        DeprecationWarning, stacklevel=2)
+
```
[the remark this answers is not in the document]
Comment (fancidev) [reply scipy/spatial/distance.py:969]: Yes it’s redundant, I forgot to remove it after writing the decorator…

File: scipy/spatial/distance.py [hunk 3]
```diff
@@ -1573,8 +1591,20 @@ def russellrao(u, v, w=None):
...
+            "The sokalmichener metric is deprecated since Scipy 1.15.0 and will "
+            "be removed in SciPy 1.17.0.", DeprecationWarning, stacklevel=2)
+        return fn(*args, **kwargs)
+    return wrapper
+
+
+@_deprecate_sokalmichener
 def sokalmichener(u, v, w=None):
     """
+    (Deprecated)
```
[the remark this answers is not in the document]
Comment (fancidev) [reply scipy/spatial/distance.py:1607]: Sure, let me do that!

File: scipy/spatial/distance.py [hunk 4]
```diff
@@ -890,9 +890,23 @@ def jaccard(u, v, w=None):
     return (a / b) if b != 0 else np.float64(0)
 
 
+def _deprecate_kulczynski1(fn):
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        warnings.warn(
+            "The kulczynski1 metric is deprecated since Scipy 1.15.0 and will "
```
[the remark this answers is not in the document]
Comment (fancidev) [reply scipy/spatial/distance.py:897]: Good catch.

File: scipy/spatial/distance.py [hunk 5]
```diff
@@ -1573,8 +1591,20 @@ def russellrao(u, v, w=None):
...
 
 
+def _deprecate_sokalmichener(fn):
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        warnings.warn(
+            "The sokalmichener metric is deprecated since Scipy 1.15.0 and will "
+            "be removed in SciPy 1.17.0.", DeprecationWarning, stacklevel=2)
+        return fn(*args, **kwargs)
+    return wrapper
```
[the remark this answers is not in the document]
Comment (fancidev) [reply scipy/spatial/distance.py:1601]: I didn’t know this internal decorator, it looks exactly what I wanted. Thanks for pointing this out! I’ll patch accordingly.

File: scipy/spatial/distance.py (hunk 1 above)
Comment (fancidev) [reply scipy/spatial/distance.py:1599]: Good suggestion. Let me add that too!

**Comment by fancidev** [comment]
Thanks for the review @tylerjereddy! It’s really helpful.

> I didn't see any comments/responses for `kulczynski1` in [gh-21187](https://github.com/scipy/scipy/issues/21187), which makes me a bit more nervous.

I posted a topic in the forum two months ago (https://discuss.scientific-python.org/t/rfc-deprecate-scipy-spatial-distance-kulczynski1/1319) but got no response either.

I’m not too worried about the lack of response — I assume that’s simply because those who look at these posts don’t use the metric, and those who use the metric don’t look at these posts. That’s why I guess the easiest way to find out is to actually deprecate and remove them :-)

Why I think kulczynski1 may not be used in the wild is the following. Suppose the metric is used for clustering analysis, which appears to be the most common use of these metrics. A clustering algorithm typically works with many metrics, so it is likely the clustering algorithm treats kulczynski1 as a dissimilarity measure. Since kulczynski1 actually returns a similarity value, I expect the clustering result to be nowhere close to make sense. So the user is likely to proceed with other metrics, making kulczynski1 not used.

A similar reasoning suggests sokalmichener is more subtle to replace. Maybe the user expects rogerstanimoto, maybe the user expects matching, or maybe the user doesn’t care. But rogerstanimoto is not substantially different from matching (they only differ in whether 0-0 count enters the denominator), so a clustering result using either metric might yield similar result. This suggests it might be difficult to observe the effect of (changing) the behavior of sokalsmichener. That’s why I suggest to deprecate it.

> I tried doing some GitHub code searching to see how popular these metrics are, but didn't reach a decisive conclusion. It may make sense to ask i.e., CJ to do a quick check on the Google code search to see if they are popular.

@perimosocordiae Do you by chance have a way to search for usage of “kulczynski1” and “sokalmichener” of `scipy.spatial.distance` by code in the wild?

> One other thing I haven't seen discussed is that `scikit-learn` also has distance computation infrastructure quite similar to ours: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.DistanceMetric.html
> 
> They have `sokalmichener` and `rogerstanimoto` both defined identically in their docs, so I'd like to see some discussion/coordination with them on:
> 
> 1. why they are doing that? (because of us?)
> 2. if they agree on the decision to have only one of them moving forward (we don't need their permission, but it might be good to just check with them)
> 
> They also have `kulsinski` defined there, so I'd again be interested in hearing their thoughts/coordinating a little on it. Maybe a quick discourse thread to coordinate a bit, or just ping them here maybe..

(TBC)

**Comment by perimosocordiae** [comment]
My code searching reveals no usage of `kulczynski1`, and some very minor usage of `sokalmichener` that would be trivial to replace.

**Comment by fancidev** [comment]
Thanks @perimosocordiae for the help! The code search results are reassuring that the impact of deprecation and removal should be rather limited and manageable.

@tylerjereddy I have applied the refactoring changes you suggested, and suggested alternatives in the deprecation message to help the user migrate. I also added deprecation messages to the doc of cdist/pdist for completeness.

**Comment by thomasjpfan** [comment]
For scikit-learn, removing `kulczynski1` should have minimal impact. As for `sokalmichener`, I opened https://github.com/scikit-learn/scikit-learn/issues/29925 to discuss in scikit-learn.

**Comment by tylerjereddy** [comment]
Thanks, let's see if anything comes up in the sklearn discussion first perhaps, then I'll circle back pretty soon.

**Comment by tylerjereddy** [comment]
Ok, I remain a bit hesitant on `kulczynski1` vs. just "fixing" it, but it looks like the community/infrastructure (`scikit-learn`, Google internal code search) is either supportive or appears to incur only minor changes to support the proposed cleanups.

Let's see if the deprecations encounter any pushback in the wild then, thanks.

**Comment by fancidev** [comment]
> Ok, I remain a bit hesitant on `kulczynski1` vs. just "fixing" it, but it looks like the community/infrastructure (`scikit-learn`, Google internal code search) is either supportive or appears to incur only minor changes to support the proposed cleanups.
> 
> Let's see if the deprecations encounter any pushback in the wild then, thanks.

Thank you @tylerjereddy!
